components/hab/src/main.rs

```diff
@@ -1348,12 +1349,26 @@ fn binlink_dest_dir_from_matches(matches: &ArgMatches) -> PathBuf {
 }
 
 fn install_sources_from_matches(matches: &ArgMatches) -> Result<Vec<InstallSource>> {
-    matches.values_of("PKG_IDENT_OR_ARTIFACT")
+    matches
+        .values_of("PKG_IDENT_OR_ARTIFACT")
         .unwrap() // Required via clap
         .map(|t| t.parse().map_err(Error::from))
         .collect()
 }
 
+fn excludes_from_matches(matches: &ArgMatches) -> Result<Vec<PackageIdent>> {
+    match matches.values_of("EXCLUDE") {
+        Some(idents) => {
+            let mut list = Vec::with_capacity(idents.len());
```

Comment:
I'm assuming the reason you didn't use `map` here  is that `PackageIdent::from_str` can fail, but `map`'s closure can not.

However, it occurs to me that it would be better UX to validate our inputs earlier. If you add a `{valid_ident}` to the definition of `@arg EXCLUDE`*, then you can be certain the of success and use the more succinct:

```rust
            Ok(idents.map(|i| PackageIdent::from_str(i).unwrap()).collect())
```

You just need to add a definition for the validator, but the hard work is done already:
```rust
fn valid_ident(val: String) -> result::Result<(), String> {
    match PackageIdent::from_str(&val) {
        Ok(_) => Ok(()),
        Err(_) => Err(format!(
            "'{}' is not valid. Package identifiers have the form origin/name[/version[/release]]",
            &val
        )),
    }
}
```

\* and `@arg PKG_IDENT` and a bunch of other args in other commands, actually

Reply:
Exactly.  We seemed to have both forms through the code depending on what work was being done.  The validator seems like a nice approach.  I'll look at that